pkg/importer/spanner/sysl.arrai

```diff
@@ -0,0 +1,140 @@
+# size returns the size of an attribute
+let size = \length
+    cond {
+        length = 'MAX': '',
+        //seq.has_prefix('0x', length): $`(${//seq.trim_prefix('0x', length)})`,
+        length > 0: $`(${length})`,
+    };
+
+# hasAttributePattern checks whether an attribute has associated patterns
+let hasAttributePattern = \entity \attr
+    cond {
+        attr.options?:{}: true,
+        (entity.primary_key where //seq.contains(attr.name, .)): true,
+        let fks = (entity.foreign_keys => ((.foreign_keys where .attribute = attr.name) =>
+            .reference_attribute orderby .)) where .; fks: true,
+        attr.length = 'MAX': true,
+        //seq.has_prefix('0x', attr.length): true,
+    };
+
+# hasEntityPattern checks whether an entity has associated patterns
+let hasEntityPattern = \entity \model
+    entity.primary_key | entity.foreign_keys | entity.cluster | model.indexes;
+
+# attributePatternDelimiter returns the token based on attribute value
+let attributePatternDelimiter = \entity \attr \token
+    cond {
+        hasAttributePattern(entity, attr): token
+    };
+
+# entityPatternDelimiter returns the token based on entity value
+let entityPatternDelimiter = \entity \model \token
+    cond {
+        hasEntityPattern(entity, model): token
+    };
+
+# sortingOrder determines and appends sorting order
+let sortingOrder = \e \attr
+    let keyOrder = //seq.split('(', //seq.join('', (e.primary_key where //seq.contains(attr.name, .)) orderby .));
+    cond {
+        keyOrder count > 1: cond keyOrder(1) {
+            'asc)': ', ~asc',
+            'desc)': ', ~desc'
+        }
+    };
```

Comment:
Rather than doing this strange split and double check, can you better express what you're actually doing? It looks like you're building a big string, then checking whether it contains `(asc)` or `(desc)`, or possibly checking which one comes first.

Perhaps:
```suggestion
    let keys = //seq.join('', e.primary_key where //seq.contains(attr.name, .) orderby .);
    cond //re.compile('\\((asc|desc)\\)').match(keys) {[[_, ord], ...]: ord rank (:.@)};
```

Reply:
Intent looks pretty clear from the code get the chosen order from primary_key attribute and return an '~asc' or `~desc` depending on the chosen orderby.

Reply:
The intent of splitting on `(` is unclear; it requires knowledge of the structure of `primary_key` (fair enough), and the fact that the sort order is encoded into the `primary_key` column name string (which it shouldn't be), and confidence in the fact that the only time `(` will exist in a column name is immediately before a sort order (which is fragile if this assumption ever changes).

For example: what if more data is encoded in the name? Or what if there are multiple columns with similar names and different sort orders? Your split for `['foo(asc)', 'foobar(desc)']` will return something like `['foo', 'asc)foobar', 'desc)']`, and keyOrder(1) will be `asc)foobar` which won't match.

What I mean by "intent" is that the implementation should closely match the conceptual logic of the function. In this case, the function is conceptually searching for either `(asc)` or `(desc)` in one of the PK column name strings (not clear which one if there are multiple). When you look at the implementation it's joining strings, splitting on `(` and picking the second element. None of that is related to the concept of searching for sort orders.

Having looked into the structure of `primary_key`, I think my suggestion could be further improved. It should literally say "check each column for an order, and return that order with a tilde in front." `//seq.contains(attr.name)` is not appropriate in this case, since it's supposed to be an exact match.

```
    let re = //re.compile($`${attr.name}\((asc|desc)\)`);
    let order = e.primary_key => (re.match(.)(0)?(1)?:{} rank (:.@)) where .;
    cond order {{x}: $`~${x}`}
```

An even better solution would be to not embed the sort order in the column name; make it a tuple with `name` and `order` instead of a string.

Reply:
fine